specification/dartLangSpec.tex

```diff
@@ -5204,28 +5206,43 @@ \subsubsection{Correct Member Overrides}
   $F$ must then be a subtype of $F'$.
 
   \commentary{%
-  The subtype requirement ensures that argument list shapes
-  that are admissible for an invocation of a method with signature $m'$
-  are also admissible for an invocation of a method with signature $m$.
-  For instance, $m'$ may accept 2 or 3 positional arguments,
-  and $m$ may accept 1, 2, 3, or 4 positional arguments, but not vice versa.
-  This is a built-in property of the function type subtype rules.
-  %
-  Note that a member signature differs from
-  an underlying syntactic declaration $D$ in a class $C$.
-  In particular,
-  a parameter in a member signature has the modifier \COVARIANT{}
-  if and only if the parameter is covariant-by-declaration
-  (\ref{covariantParameters}),
-  and that may be the case due to declarations in a supertype of $C$,
-  so that modifier need not be present in $D$.
-  %
-  There is an additional potential compile-time error associated with
-  a parameter which is covariant-by-declaration
-  (\ref{instanceMethods}).
-  But we cannot cover that here as a property of member overrides,
-  because it is concerned with declarations in all superinterfaces,
-  indirect as well as direct.%
+    The subtype requirement ensures that argument list shapes
+    that are admissible for an invocation of a method with signature $m'$
+    are also admissible for an invocation of a method with signature $m$.
+    For instance, $m'$ may accept 2 or 3 positional arguments,
+    and $m$ may accept 1, 2, 3, or 4 positional arguments, but not vice versa.
+    This is a built-in property of the function type subtype rules.
+
+    Note that a member signature differs from
+    an underlying syntactic declaration $D$ in a class $C$.
+    In particular,
+    a parameter in a member signature has the modifier \COVARIANT{}
+    if and only if the parameter is covariant-by-declaration
+    (\ref{covariantParameters}),
+    and that may be the case due to declarations in a supertype of $C$,
+    so that modifier need not be present syntactically in $D$.
+  }
+
+  If $p'$ is a formal parameter in $m'$ that has the modifier \COVARIANT,
+  the corresponding parameter $p$ in $m$ must also have the modifier \COVARIANT,
+  unless the type of $p$ is a top type
+  (\ref{superBoundedTypes}).
+
+  \commentary{%
+    This can only occur in the case where $m$ is the signature of
+    a concrete member declaration inherited from a superclass.
+    Otherwise, if $m$ needs to be a correct override of $m'$ then
+    $m'$ is already a member of a superinterface of the enclosing class
+    of the declaration whose signature is $m$.
+    Every \COVARIANT{} parameter of $m'$ is then also \COVARIANT{} in $m$,
+    with or without an explicit occurrence of the modifier \COVARIANT.
+
+    There is one additional potential compile-time error associated with
```

Comment:
Again with the compile-time errors in a predicate :)

I'd just drop this. It's not relevant to the *definition* of when something is a valid override of something else.

Reply:
I really think it's important to mention that there is a check which is performed relative to all superinterfaces, and also that it can't be a property of 'correct member override' because we want to make that property specific to the relation to a single superinterface.

I moved this paragraph out of the bullet list, such that it is a comment on the section as a whole rather than a comment on one of the cases in the predicate. I think it's more readable now.